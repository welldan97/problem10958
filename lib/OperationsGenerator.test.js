const OperationsGenerator = require('./OperationsGenerator');
const { add, subtract } = require('../specifics/operations');

const present = ({ value: { operations, priorities } }) => ({
  operations,
  priorities,
});

describe('OperationsGenerator', () => {
  describe('without batches', () => {
    it('generates operations by one', () => {
      const numbers = [1, 2, 3];
      const operations = [add, subtract];
      const operationsGenerator = new OperationsGenerator(numbers, operations);
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(false);
      operationsGenerator.next();
      expect(present(operationsGenerator)).toMatchSnapshot();
      expect(operationsGenerator.done).toEqual(true);
    });
  });

  describe('with  batches', () => {
    it('generates operations by one for given batch', () => {
      const numbers = [1, 2, 3];
      const operations = [add, subtract];
      const operationsGenerator1 = new OperationsGenerator(
        numbers,
        operations,
        {
          start: 0,
          length: 2,
        },
      );

      expect(present(operationsGenerator1)).toMatchSnapshot();
      expect(operationsGenerator1.done).toEqual(false);
      operationsGenerator1.next();
      expect(present(operationsGenerator1)).toMatchSnapshot();
      expect(operationsGenerator1.done).toEqual(true);

      const operationsGenerator2 = new OperationsGenerator(
        numbers,
        operations,
        {
          start: 2,
          length: 2,
        },
      );

      expect(present(operationsGenerator2)).toMatchSnapshot();
      expect(operationsGenerator2.done).toEqual(false);
      operationsGenerator2.next();
      expect(present(operationsGenerator2)).toMatchSnapshot();
      expect(operationsGenerator2.done).toEqual(true);

      const operationsGenerator3 = new OperationsGenerator(
        numbers,
        operations,
        {
          start: 6,
          length: 2,
        },
      );

      expect(present(operationsGenerator3)).toMatchSnapshot();
      expect(operationsGenerator3.done).toEqual(false);
      operationsGenerator3.next();
      expect(present(operationsGenerator3)).toMatchSnapshot();
      expect(operationsGenerator3.done).toEqual(true);
    });
  });

  it('calculates total number of combinations', () => {
    const length = OperationsGenerator.calculateLength(3, 4);
    expect(length).toEqual(32);
  });
});
