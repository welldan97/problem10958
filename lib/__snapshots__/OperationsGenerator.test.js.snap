// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationsGenerator with  batches generates operations by one for given batch 1`] = `
Object {
  "operations": Uint8Array [
    0,
    0,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator with  batches generates operations by one for given batch 2`] = `
Object {
  "operations": Uint8Array [
    0,
    0,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator with  batches generates operations by one for given batch 3`] = `
Object {
  "operations": Uint8Array [
    0,
    1,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator with  batches generates operations by one for given batch 4`] = `
Object {
  "operations": Uint8Array [
    0,
    1,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator with  batches generates operations by one for given batch 5`] = `
Object {
  "operations": Uint8Array [
    1,
    1,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator with  batches generates operations by one for given batch 6`] = `
Object {
  "operations": Uint8Array [
    1,
    1,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 1`] = `
Object {
  "operations": Uint8Array [
    0,
    0,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 2`] = `
Object {
  "operations": Uint8Array [
    0,
    0,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 3`] = `
Object {
  "operations": Uint8Array [
    0,
    1,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 4`] = `
Object {
  "operations": Uint8Array [
    0,
    1,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 5`] = `
Object {
  "operations": Uint8Array [
    1,
    0,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 6`] = `
Object {
  "operations": Uint8Array [
    1,
    0,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 7`] = `
Object {
  "operations": Uint8Array [
    1,
    1,
  ],
  "priorities": Uint8Array [
    0,
    1,
  ],
}
`;

exports[`OperationsGenerator without batches generates operations by one 8`] = `
Object {
  "operations": Uint8Array [
    1,
    1,
  ],
  "priorities": Uint8Array [
    1,
    0,
  ],
}
`;
